Pull Request: chore/generalize screen architecture

tombu opened the pull request:
Like discussed in issue #113 the four screens have some duplicate code. To be more general concerning the views I added a `Appendable` mixin.

The mixin has a `appendView()` and a `removeView()` method that append/remove the view which is also a part of the mixin but has to be set by the particular screen.
In addition the mixin overrides the `destroy()` method so the views are removed automatically when the screen gets destroyed.

**Other stuff:**
- I left `removeView()` public so that views can also be removed by hand. I´m not sure if we ever need it ... so maybe it should be private
- As you will recognize every screen has a private `_appendScreen()` method. All those look pretty identical. Instead we could extend the `appendView()` method of `Appendable` so that it takes a view and cares for setting the `view` property and appending the view.

How do you think about it?

tombu committed:
adds Appendable mixin and spec file

tombu committed:
adds 'appendView' and 'removeView' to Appendable

tombu committed:
implement Appendable in existing screens

tombu committed:
extracts destroy process from every screen and moves it into Appendable

tombu committed:
fixes failing test

stravid commented on spec/javascripts/unit/slotcars/shared/lib/appendable_spec.js.coffee:
```diff
@@ -0,0 +1,42 @@
+    @appendable.view = @viewMock
+
+  it 'should require a view', ->
+    (expect => Appendable.apply {}).toThrow()
```
I understand what you want to test, but I don't understand what this line does. Could you please explain? :)

tombu replied:
I´m aware that this is not obvious.
This line applies the Appendable mixin on an object and assumes an error when 'view' property is not set.
`Ember.required()` just works when the mixin is applied after creation which is not surprising because before creation it can´t be detected if a property is set.

Instead of `{}` you could of course use `Ember.Object.create()`.

Maybe I should just remove this test?! The only thing it really tests is that the value of `view` is `Ember.required()` and not just `null`.

stravid replied:
Why is there a fat arrow?

tombu replied:
It´s an anonymous function that calls `Appendable.apply {}`. We already used that in some other tests. It´s necessary do recognize errors that come from Ember itself. @DominikGuzei can presumably teach you the deeper logic behind this. I can´t :(

tombu committed:
generalizes even more

removes 'appendToApplication()' from all screens, appending is handled by 'Appendable' mixin

tombu committed:
integrates new append-functionality

tombu committed:
fixes bug when leaving build screen

stravid commented:
When should we merge?

Commit:
merges development into chore/generalize-screen-architecture

tombu committed:
fixes failing test in play screen
